### ISSUE TITLE
How to serialize/deserialize solver

### ISSUE THREAD

**Issue by mwpowellhtx** [google/or-tools#325]
I am interested to save and restore a solver between sessions. Is it possible? I am using the .NET package. Thanks!

**Comment by mwpowellhtx** [comment]
How to do this?

**Comment by mwpowellhtx** [comment]
@lperron Would you mind clarifying?

I am reviewing a patch someone forwarded to me, which appears to be adapted almost verbatim from a [Run example](http://github.com/google/or-tools/blob/3d765b4f4adf48fada92d2f84a834c3ce553d9bf/examples/cpp/solve.cc) . AFAIK, this is a good start for Linear Solver serialization.

I'll start with the Linear Solver since the examples seem to be more readily available, but ultimately I am more interested in Constraint Solver since that is much of the work I am doing.

For Constraint Solver, initial analysis leads me to think that some sort of wrapper function around ExportModel and LoadModel would be desirable; i.e. export model to protobuf, save protobuf to file; followed by load protobuf from file, load model from protobuf.

Thanks...

**Comment by mwpowellhtx** [comment]
@lperron Couple of questions following up:

* Where is ``FLAGS_forced_mps_format`` defined? I take it as a ``std::string``, but where? Third party library?

* Also, studying the [example](http://github.com/google/or-tools/blob/3d765b4f4adf48fada92d2f84a834c3ce553d9bf/examples/cpp/solve.cc), how does the model proto actually land in the solver? I see a reference to a ``.name()`` call, but what does a "name" have to do with the model? variables? constraints or other rules? etc?

Thanks!

**Comment by mwpowellhtx** [comment]
@lperron I have a couple more follow on questions. Speaking about Linear Solver for the moment:

* Let's say that model and/or request proto, once parsed, have a time limit set. How is that conveyed to the solver? It seems like this is set as a global, once, and is not part of the model itself. Which strikes me as a bit of a snafu in the making, but I could be wrong. Is there a better way to handle saving/loading the time limit value?

It seems to me that proper support for time limit is not going to be easy. Once it is set, the model and/or the time limit can easily get away from you. A bit deeper introspection, I don't want to turn time limit treatment into a huge rabbit hole for purposes of satisfactory serialization.

Thoughts?

**Comment by mwpowellhtx** [comment]
@lperron Approaching an initial commit and merge candidate in order to expose a Linear Solver ReadModelFromFile patch.

```
ws2_32.lib(WS2_32.dll) : warning LNK4006: __NULL_IMPORT_DESCRIPTOR already defined in psapi.lib(PSAPI.DLL); second definition ignored
ws2_32.lib(WS2_32.dll) : warning LNK4221: This object file does not define any previously undefined public symbols, so it will not be used by any link operation that consumes this library
shlwapi.lib(SHLWAPI.dll) : warning LNK4006: __NULL_IMPORT_DESCRIPTOR already defined in psapi.lib(PSAPI.DLL); second definition ignored
shlwapi.lib(SHLWAPI.dll) : warning LNK4221: This object file does not define any previously undefined public symbols, so it will not be used by any link operation that consumes this library
sulum_interface.obj : warning LNK4221: This object file does not define any previously undefined public symbols, so it will not be used by any link operation that consumes this library
cplex_interface.obj : warning LNK4221: This object file does not define any previously undefined public symbols, so it will not be used by any link operation that consumes this library
```

There are a couple of linker warnings, probably should be errors. I hadn't noticed them before, but I also was not looking for them. The DLL does get linked it would seem.

**Comment by mwpowellhtx** [comment]
@lperron Before I commit, here's the code I've got thus far. Basically adapted from *examples/cpp/solver.cc*.

```C++
bool MPSolver::ReadModelFromFile(const std::string& model_path, MpsReadMode read_mode) {
  MPModelProto model_proto;
  // TODO: TBD: case may be important here, either in the context of this function,
  // or more broadly in HasSuffixString and HasPrefixString ...
  if (HasSuffixString(model_path, ".mps") ||
    HasSuffixString(model_path, ".mps.gz")) {
    // TODO: TBD: is it really necessary to duplicate concerns, and potentially read cycles, due to an overlapping mode?
    glop::LinearProgram linear_program_free;
    glop::LinearProgram linear_program_fixed;
    glop::MPSReader mps_reader;

    auto should_read_free = (read_mode & MPS_FREE_FORM) == MPS_FREE_FORM;
    auto should_read_fixed = (read_mode & MPS_FIXED_FORM) == MPS_FIXED_FORM;

    const auto did_read_free = should_read_free &&
      mps_reader.LoadFileWithMode(model_path, true, &linear_program_free);
    auto did_read_fixed = should_read_fixed &&
      mps_reader.LoadFileWithMode(model_path, false, &linear_program_fixed);

    CHECK(did_read_free || did_read_fixed)
      << "Failed to parse the mps file '" + model_path + "'.";

    if (did_read_free && did_read_fixed) {
      // Fall back on the Free form mode.
      did_read_fixed = false;
    }

    if (did_read_free) {
      LOG(INFO) << "Read file '" + model_path + "' in free format.";
      LinearProgramToMPModelProto(linear_program_free, &model_proto);
    } else if (did_read_fixed) {
      LOG(INFO) << "Read file '" + model_path + "' in fixed format.";
      LinearProgramToMPModelProto(linear_program_fixed, &model_proto);
    }
  } else {
    MPModelRequest request_proto;

    CHECK(ReadFileToProto(model_path, &model_proto))
      << "Failed to read file '" + model_path + "' into model proto.";
    CHECK(ReadFileToProto(model_path, &request_proto))
      << "Failed to read file '" + model_path + "' into request proto.";

      // If the input proto is in binary format, both ReadFileToProto could return true.
    // Instead use the actual number of variables found to test the correct format of
    // the input.
    const auto is_model_proto = model_proto.variable_size() > 0;
    const auto is_request_proto = request_proto.model().variable_size() > 0;
    // If neither is the case then log a fatal error.
    if (!(is_model_proto || is_request_proto)) {
      LOG(FATAL) << "Failed to parse '" << model_path
        << "' as an MPModelProto or an MPModelRequest.";
      return false;
    } else {
      // TODO: TBD: not sure if this completely guards against the BOTH being the case.
      // What happens with is_model_proto, for instance?

      // We have checked whether neither case exists; also rule out both cases.
      CHECK(!(is_model_proto && is_request_proto));

      // TODO: TBD: how good/better/best to relay and/or set the Linear Solver time limit?
      if (is_request_proto) {
        LOG(INFO) << "Read input proto as an MPModelRequest.";

        // TODO: TBD: does the timer need to be re-factored to the model itself?
        if (request_proto.has_solver_time_limit_seconds()) {
          set_time_limit(static_cast<int64>(request_proto.solver_time_limit_seconds() * 1000));
        }

        // Run with the request proto in lieu of the "model proto".
        model_proto.Swap(request_proto.mutable_model());
      }
      else {
        LOG(INFO)
          << "Read input proto as an MPModelProto.";
      }

      std::string error_message;

      // Load the proto into the solver.
      const MPSolverResponseStatus status =
        LoadModelFromProto(model_proto, &error_message);

      CHECK_EQ(MPSOLVER_MODEL_IS_VALID, status)
        << MPSolverResponseStatus_Name(status) << ": " << error_message;

      LOG(INFO)
        << "Solver Dimension: " << NumConstraints() << " x " << NumVariables();
    }
  }

  return true;
}
```

And which SWIG exposes as:

```C#
public bool ReadModelFromFile(string model_name, int read_mode)
{
    // ...
}
```

But for the linker warning, seems to be okay in the build itself. Haven't verified in a run-time yet.

**Comment by mwpowellhtx** [comment]
@lperron Slightly refined. Verified build, but untested.

```C++
bool MPSolver::ReadModelFromFile(const std::string& model_path, MpsReadMode read_mode) {
  MPModelProto model_proto;
  // TODO: TBD: case may be important here, either in the context of this function,
  // or more broadly in HasSuffixString and HasPrefixString ...
  if (HasSuffixString(model_path, ".mps") || // TODO: TBD: this is reading from linear program (?): shouldn't this be checking for LP then?
    HasSuffixString(model_path, ".mps.gz")) {
    // TODO: TBD: is it really necessary to duplicate concerns, and potentially read cycles, due to an overlapping mode?
    glop::LinearProgram linear_program_free;
    glop::LinearProgram linear_program_fixed;
    glop::MPSReader mps_reader;

    auto should_read_free = (read_mode & MPS_FREE_FORM) > 0;
    auto should_read_fixed = (read_mode & MPS_FIXED_FORM) > 0;

    const auto did_read_free = should_read_free &&
      mps_reader.LoadFileWithMode(model_path, true, &linear_program_free);
    auto did_read_fixed = should_read_fixed &&
      mps_reader.LoadFileWithMode(model_path, false, &linear_program_fixed);

    CHECK(did_read_free || did_read_fixed)
      << "Failed to parse the mps file '" + model_path + "'.";

    if (did_read_free && did_read_fixed) {
      // Fall back on the Free form mode.
      did_read_fixed = false;
    }

    if (did_read_free) {
      LOG(INFO) << "Read file '" + model_path + "' in free format.";
      LinearProgramToMPModelProto(linear_program_free, &model_proto);
    } else if (did_read_fixed) {
      LOG(INFO) << "Read file '" + model_path + "' in fixed format.";
      LinearProgramToMPModelProto(linear_program_fixed, &model_proto);
    }
  } else { // Read from .lp (?)
    MPModelRequest request_proto;

    // // TODO: TBD: actually, on second thought, maybe these two SHOULD NOT be checked, after all...
    // CHECK(ReadFileToProto(model_path, &model_proto))
    //   << "Failed to read file '" + model_path + "' into model proto.";
    // CHECK(ReadFileToProto(model_path, &request_proto))
    //   << "Failed to read file '" + model_path + "' into request proto.";
    ReadFileToProto(model_path, &model_proto);
    ReadFileToProto(model_path, &request_proto);

    // If the input proto is in binary format, both ReadFileToProto could return true.
    // Instead use the actual number of variables found to test the correct format of
    // the input.
    const auto is_model_proto = model_proto.variable_size() > 0;
    const auto is_request_proto = request_proto.model().variable_size() > 0;
    // If neither is the case then log a fatal error.
    if (!(is_model_proto || is_request_proto)) {
      LOG(FATAL) << "Failed to parse '" << model_path
        << "' input proto as either MPModelProto or MPModelRequest.";
      return false;
    } else {
      // TODO: TBD: not sure if this completely guards against the BOTH being the case.
      // What happens with is_model_proto, for instance?

      // We have checked whether neither case exists; also rule out both cases.
      CHECK(!(is_model_proto && is_request_proto));

      // TODO: TBD: how good/better/best to relay and/or set the Linear Solver time limit?
      if (is_request_proto) {
        LOG(INFO) << "Read input proto as an MPModelRequest.";

        // TODO: TBD: could potentially be refactored if, for instance, solver time limit is swapped into model_proto above...
        if (request_proto.has_solver_time_limit_seconds()) {
          set_time_limit(static_cast<int64>(request_proto.solver_time_limit_seconds() * 1000));
        }

      // Run with the request proto in lieu of the model proto.
        model_proto.Swap(request_proto.mutable_model());
      }
      else {
        LOG(INFO) << "Read input proto as an MPModelProto.";

        // TODO: TBD: what happens, if anything, for solver time limit as part of MPModelProto (?), esp. since not a member...
      }

      std::string error_message;

      // TODO: TBD: here, we aren't concerned with previous solver state? i.e. clearing it out at all?
      //  i.e. likely usage scenario, starting from new solver instance, clean slate...

      // Load the model proto into the solver.
      const MPSolverResponseStatus status =
        LoadModelFromProto(model_proto, &error_message);

      CHECK_EQ(MPSOLVER_MODEL_IS_VALID, status)
        << MPSolverResponseStatus_Name(status) << ": " << error_message;

      LOG(INFO) << "Solver Dimension: " << NumConstraints() << " x " << NumVariables();
    }
  }

  return true;
}
```

**Comment by mwpowellhtx** [comment]
@lperron Couple of assumptions I am making based on what's there:

* Makes an appropriate read of either ``MPS`` (or ``gzipped MPS``), versus LP format in the *else* clause.

* Making appropriate juggling of either *fixed* and/or *free form* input during the ``MPS`` read.

* The transmission of the *solver time limit* seems somewhat limited to me. That's a rabbit hole I'm not jumping down just now as the implications seem significantly more vast to me than the issue warrants.

* Assumes that the current state of the solver can simply be loaded, and that nothing needs to be cleared from the solver prior to load.

**Comment by mwpowellhtx** [comment]
@lperron Which will also require a couple of minor changes in the ``HasSuffixString`` and ``HasPrefixString`` string utilities.

Can go about this a hundred different ways, probably. I saw one example leveraging the ``std::basic_string<>`` character traits for case sensitivity. However, given the mixed platform support, this approach is about as good as any.

```C++
#include <algorithm>
```

Defaulting to case sensitivity for **nix* purposes. Otherwise, can specific insensitivity, i.e. for Windows.

```C++
inline bool HasSuffixString(const StringPiece& str, const StringPiece& suffix, bool case_insensitive = false) {

  // Just return when case sensitive.
  if (!case_insensitive) {
    return str.ends_with(suffix);
  }

  return str.length() > suffix.length() &&
         std::equal(str.begin() + (str.length() - suffix.length()), str.end(), suffix.begin(),
         [](const char& a, const char& b) { return ::tolower(a) == ::tolower(b); });
}
```

```C++
inline bool HasPrefixString(const StringPiece& str, const StringPiece& prefix, bool case_insensitive = false) {

  // Just return when case sensitive.
  if (!case_insensitive) {
    return str.starts_with(prefix);
  }

  return str.length() > prefix.length() &&
         std::equal(str.begin(), str.begin() + prefix.length(), prefix.begin(),
         [](const char& a, const char& b) { return ::tolower(a) == ::tolower(b); });
}
```

I'm pretty sure the iterator boundaries are correct, but please double check them.

**Comment by lperron** [comment]
Hi Michael,

I wonder what is the end goal of this. If you want to store models, or send
models from masters to workers, you can use MPS or proto file.
If you do not need MPS, I would stick to proto files. Furthermore, we have
code to make them work in all languages.

We have implemented a linear solver proto that allows these forms of
communications.
You create a MPSolverRequest -- which does contains time limit -- and then
solve it with MPSolver::SolveWithProto().

And you can store these proto, send them in binary form stored inside a
string.

For CP, we have a similar mechanism with the 'model.proto' file. And we are
working on the same features with our CP on SAT solver.

So bottom line,

I you do not need MPS, stick to proto.
If you need MPS, you do not need linear_solver, just use the solver
binaries.

I hope this helps.

--Laurent

Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00


2017-03-15 20:49 GMT+01:00 Michael Powell <notifications@github.com>:

**Comment by mwpowellhtx** [comment]
@lperron Indeed, well, among my goals is to serialize the solver state, including variables, constraints, and anything else, such as already visited solution candidates (i.e. decision builder?), necessary to restore and continue running where it left off. That's one of my goals for the Constraint Solver. Possibly for the Linear Solver, eventually, but I don't have that use case, yet.

``SolveWithProto`` is not (currently?) exposed via the SWIG to ``Solver``. From the C++ perspective, a *const pointer* is probably the right thing to do. But it isn't being interpreted correctly by the SWIG generated code. Either ``ExportModel`` needs to return a ``CpModel``, or perhaps accept an ``out CpModel proto``.

Okay, so taking a look at the SWIG-generated assembly.

Further, from the C# perspective, there is ``Solver.ExportModel`` and ``Solver.LoadModel``, correct? I'm not sure how ``SWIGTYPE_p_operations_research__CpModel`` (brilliant name, BTW) can be created since the constructors are not visible to outside callers, and it's not like there is a ``ref`` or an ``out`` parameter, from C# perspective.

Perhaps, notwithstanding some further C++ and/or SWIG work needing to be done, the first thing I want to do is extend ``SWIGTYPE_p_operations_research__CpModel`` to something more meaningfully named, like simply ``class CpModel : SWIGTYPE_p_operations_research__CpModel {}``.

Minimally, do the ``SWIGTYPE_p_operations_research__CpModel`` constructor(s) need to be exposed?

**Comment by mwpowellhtx** [comment]
@lperron So when I try something like this:

```C#
public class CpModel : SWIGTYPE_p_operations_research__CpModel { }
```

```C#
[Test]
public void Testing()
{
    var s = new Solver("Test");
    var d = new CpModel();
    var bit = s.MakeIntVar(0, 1);
    bit.SetName("Bit");
    try
    {
        s.ExportModel(d);
    }
    catch (Exception ex)
    {
    }
    var s2 = new Solver("");
    s2.LoadModel(d);
}
```

I get as far as ``ExportModel`` and the whole thing crashes and burns. No exception. Just *CRASH*.

Without diving further into the ``ExportModel`` code itself, it seems to me the fact that the ``CpModel`` is, but default basically a proxy in a null state internally (?), that the whole thing dies.

Which is my understanding of the IntPtr.Zero, it's like having a null C++ pointer, handle, or whatever else it is in which it is standing.

```C#
protected SWIGTYPE_p_operations_research__CpModel()
{
    swigCPtr = new global::System.Runtime.InteropServices.HandleRef(null, global::System.IntPtr.Zero);
}
```

But I could be wrong about that.

Then there is the ``cPtr``, ``futureUse`` ctor. Is this not fully implemented, then? If so, what can I do to help that along?

```C#
internal SWIGTYPE_p_operations_research__CpModel(global::System.IntPtr cPtr, bool futureUse)
{
    swigCPtr = new global::System.Runtime.InteropServices.HandleRef(this, cPtr);
}
```

**Comment by mwpowellhtx** [comment]
@lperron Thoughts, comments, concerns?

Digging in a bit further, I believe there are SWIG ways of exposing a C# ``out`` parameter, leaving aside other targets for the time being. I don't care about Java or Python for these purpose, not to mention SWIG mapping is likely very different.

The ``Solver::ExportModel`` is overloaded. Leaving aside the overloads, the parameter I might be interested to overload is ``CpModel* const model_proto``.

```C++
void Solver::ExportModel(const std::vector<SearchMonitor*>& monitors,
                         CpModel* const model_proto,
                         DecisionBuilder* const db) const;

void Solver::ExportModel(const std::vector<SearchMonitor*>& monitors,
                         CpModel* const model_proto) const;

void Solver::ExportModel(CpModel* const model_proto) const;
```

Which gets mapped as this in the current release:

```C#
  public void ExportModel(SWIGTYPE_p_operations_research__CpModel proto);

  public void ExportModel(SearchMonitorVector monitors, SWIGTYPE_p_operations_research__CpModel proto);

  public void ExportModel(SearchMonitorVector monitors, SWIGTYPE_p_operations_research__CpModel proto, DecisionBuilder db);
```

Which assumes I can create a new instance of ``SWIGTYPE_p_operations_research__CpModel``, which I cannot in the current release.

For brevity I will use the simplest ``ExportModel`` overload. I'm thinking the C++ needs to change as such:

```C++
void Solver::ExportModel(CpModel* model_proto) const;
```

And the SWIG would look something like this, methinks:

```SWIG
%apply CpModel* OUTPUT { CpModel* model_proto };
  void ExportModel(CpModel* model_proto);
```

Which if I understand the SWIG bits, should yield this:

```C#
void ExportModel(out SWIGTYPE_p_operations_research__CpModel proto)
{
    // ...
}
```

Or like this:

```C++
CpModel* Solver::ExportModel() const;
```

Which I think would just return the instance outright:

```C#
SWIGTYPE_p_operations_research__CpModel ExportModel()
{
    // ...
}
```

That's the export side of things. Of course, what to do for the load/import side of things?

I'm going to poke around a little, but interested in receiving some feedback along these lines in order to get it resolved.

**Comment by mwpowellhtx** [comment]
@lperron There's also the matter of how the SWIG treatment of the *Protobuf* ``::google::protobuf::Message`` happens. It seems to me the *message-iness, so to speak, gets lost in translation. Perhaps it is not important for sake of the SWIG translation, unless we discover that it is advantageous to treat things of type *Protobuf* *Message* in a consistent manner. Never minding the fact that a public ctor is not exposed, so somehow at minimum at least internal treatment needs to happen.

**Comment by mwpowellhtx** [comment]
@lperron Supposedly this would work, at least according to the [Using typemaps](http://www.swig.org/Doc3.0/SWIGDocumentation.html#Lua_nn26) documentation, unless there is something masking the "out" keyword...

```SWIG
%include <typemaps.i>
```

Followed by this when it is important:

```SWIG
%apply CpModel *OUTPUT { CpModel *proto };
Solver::ExportModel(CpModel* proto) const;
```

It will also be important to reorder the ``CpModel* proto`` as the last parameter in the list on account of the C# language ``out`` parameter rules. That's crucial for the ``DecisionBuilder`` overload.

```C++
void ExportModel(const std::vector<SearchMonitor*>& monitors,
                 DecisionBuilder* const db, CpModel* proto) const;
```

However, I try this mapping application and does not seem to have the desired effect, or as I mentioned the ``out`` keyword is somehow otherwise being masked.

The *Solver.cs* is still reflecting this:

```C#
public void ExportModel(SWIGTYPE_p_operations_research__CpModel proto)
{
    operations_research_constraint_solverPINVOKE.Solver_ExportModel__SWIG_0(swigCPtr, SWIGTYPE_p_operations_research__CpModel.getCPtr(proto));
}
```

Next visit will be to the *Makefiles*, methinks. Perhaps there is a SWIG option that is forcing the issue one way or another.

**Comment by mwpowellhtx** [comment]
@lperron Reviewing the supporting SWIG files. I'm not sure why this is being done, particularly in a "shared" include across potentially differing language capabilities.

```SWIG
%typemap(out) std::string
%{ $result = jenv->NewStringUTF($1.c_str()); %}
```

But for language specific blocks such as the following, i.e.:

```SWIG
#ifdef SWIGJAVA
// ...
#endif
```

This could really be re-factored to separate SWIG files to help with through and through clarity.

**Comment by mwpowellhtx** [comment]
@lperron Any thoughts as to upgrading the version of SWIG dependency? Currently version *swigwin-3.0.10.zip* however [3.0.12](http://prdownloads.sourceforge.net/swig/swigwin-3.0.12.zip) is available.

**Comment by mwpowellhtx** [comment]
@lperron I noticed I was working with *python\constraint_solver.swig*. Reverted that and added the parts I figured were important in the *csharp\constraint_solver.swig*, but still not the desired effect.

At the moment, it's a delicate operation to say the least, wrapping the call in an appropriate invocation:

```SWIG
namespace operations_research {

class CpModel;

%typemap(cstype) const CpModel* "$csclassname"
%typemap(csin, 
         pre="  $csclassname temp$csinput = new $csclassname();"
        ) const CpModel* "$csclassname.getCPtr(temp$csinput)"

%typemap(cstype) CpModel* "out $csclassname"
%typemap(csin, 
         pre="  $csclassname temp$csinput = new $csclassname();", 
         post="  $csinput = new $csclassname(temp$csinput, false);", 
         cshin="out $csinput"
        ) CpModel* "$csclassname.getCPtr(temp$csinput)"

} // namespace operations_research
```

Far from ready right at the moment, but [it's a start](http://stackoverflow.com/questions/21849198/swig-objects-of-a-custom-class-as-output-argument-with-python), I think.

**Comment by mwpowellhtx** [comment]
@lperron I'm starting to wonder if such a direct path to exported and re-loaded models is not the right answer. Not a wrong answer, per se, but whether there is a better answer. Sorry if my technical vocabulary and/or grammar isn't conveying my thoughts correctly here. It is not quite second hand yet, so all the SWIG and bouncing between languages is a bit clumsy still.

This is a bit of a stumbling block for me. I see how the ``model.proto`` and so on are being gen'ed up. The ``CpModel`` message, in this case, lands as a fully featured, sort of boilerplate class, complete with strongly typed message fields. All well and good so far.

Then there is the ``SWIGTYPE_p_operations_research__CpModel`` SWIG adapter, which I really do not get at all. There is no mention of anything to do with the underlying ``CpModel``, per se, not even a PInvoke of any variety, which I would half expect. Only a loose ``System.IntPtr`` pointer variable, for instance. In the baseline, at any rate, ``SWIGTYPE_p_operations_research__CpModel`` stands in to capture the reference to a presumably PInvoke-captured C++ language level pointer, and that's it. There is no other point of origin for the "instance" from the C# or .NET perspective. All well and good, fair enough.

It seems like to me there should be some way of capturing a fresh instance of the protobuf model. Like either return a new instance, like *MakeModelProto*, something like that, exposed by the Solver, which seems like would do the right thing from a C++ language level and SWIG perspective, or be able to create a new instance of ``SWIGTYPE_p_operations_research__CpModel`` outright. But then there is a lot of nastiness of how to inform the C# type of the appropriate PInvokable C++ bits. Perhaps the *MakeModelProto* would be the simplest answer, and perhaps several other bits supporting the actual saving/loading of the message as necessary.

Then there is the internal *FLAGS* construct going on: specifically, *FLAGS_cp_export_file*. This is currently set once, that I know of. Perhaps an export file could be specified, something like:

```C#
using (var s = new Solver())
{
    s.Parameters().ExportFile = pathToExportFile;
}
```

It seems  to me that ``ExportModel`` being called with every ``ProcessConstraints`` is a bit excessive. I have in mind to run a solver during a Windows Service session, for instance. Perhaps periodic exportation would be desirable, as well as just prior to stopping the service. But not much more frequent than that, I think.

But then, I'm not sure why LoadModel is not exposed. For instance, when the service starts. Load the most recent state, reconnect with variables, etc, and proceed from there.

Hope that makes sense... Thanks again.

**Comment by mwpowellhtx** [comment]
@lperron I will close this one as it is fairly broad even for my purposes right at the moment. Will open new ones that are more specific.

**Comment by yingzong** [comment]
> If you want to store models, or send models from masters to workers, you can use MPS or proto file. If you do not need MPS, I would stick to proto files. Furthermore, we have code to make them work in all languages.

@lperron Where is the code to obtain a proto from Python? I only see `ExportModelAsMpsFormat` and `ExportModelAsLpFormat` (both of which are incomplete serializations) in the SWIG definition, and `ExportModelToProto` is not wrapped and therefore not available. So from Python the only serializations are incomplete. Furthermore there is no deserialization on the solution side.

**Comment by lperron** [comment]
You are right, the python interface for the linear solver to import/export
from/to a proto is just broken.

We will try to correct this in the next release.

Thanks

Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00
